src/shims/unix/foreign_items.rs

```diff
@@ -440,13 +440,34 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             // Miscellaneous
             "isatty" => {
                 let [fd] = this.check_shim(abi, Abi::C { unwind: false }, link_name, args)?;
-                this.read_scalar(fd)?.to_i32()?;
-                // "returns 1 if fd is an open file descriptor referring to a terminal; otherwise 0 is returned, and errno is set to indicate the error"
-                // FIXME: we just say nothing is a terminal.
-                let enotty = this.eval_libc("ENOTTY")?;
-                this.set_last_error(enotty)?;
-                this.write_null(dest)?;
-            }
+                #[cfg_attr(not(unix), allow(unused))]
+                let fd = this.read_scalar(fd)?.to_i32()?;
+                #[cfg(unix)]
+                {
+                    // "returns 1 if fd is an open file descriptor referring to a terminal;
+                    // otherwise 0 is returned, and errno is set to indicate the error"
+                    let is_tty = unsafe {
+                        libc::isatty(fd)
```

Comment:
This is not right. You'll need to coordinate with the logic in `fs.rs`. The `fd` here is a "Miri logical FD", not an actual FD on the underlying system.

Reply:
I don't think this is correct actually. The code works, and `cargo miri test` is passing a 1 into here. I think this is because nobody calls `open` then asks if the fd that comes back is a TTY, they pass a constant `0`, `1`, or `2` to this function.

Reply:
If I try to add this code which is done elsewhere in `fs.rs`:
```rust
            if let Some(file_descriptor) = this.machine.file_handler.handles.get(&fd) {
                let FileHandle { file, writable } = file_descriptor.as_file_handle()?;
```
I get this error at runtime:
```
error: unsupported operation: stdout cannot be used as FileHandle
 --> /home/ben/.rustup/toolchains/miri/lib/rustlib/src/rust/library/test/src/helpers/isatty.rs:7:22
  |
7 |             unsafe { libc::isatty(libc::STDOUT_FILENO) != 0 }
  |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ stdout cannot be used as FileHandle
  |
  = help: this is likely not a bug in the program; it indicates that the program performed an operation that the interpreter does not support
```

Reply:
The code happens to work because the FD `1`  refers to the same thing in Miri's FD table and the OS FD table. But that is not a given for all FDs. Even stdout can be `dup`ed to other FD numbers.

Reply:
Yeah, there is no [file](https://doc.rust-lang.org/nightly/std/fs/struct.File.html) handle for the "primitive" FDs.

You'll have to add a new method to the `FileDescriptor` trait, something like `as_unix_host_fd` or so. And then for stdin/stdout/stderr use `libc::STDIN_FILENO` etc.

Reply:
I gave this a shot. This whole system feels very awkward to me so I'm not sure if I did what you're looking for